hs628_tools/feature_missing.py

```diff
@@ -0,0 +1,200 @@
+    """
+
+    # Calculating the fraction of missing values in each column
+    missing_series = (len(data) - np.sum(data == data, axis=0)) / float(len(data))  # type: ndarray
+
+    # Find the columns with a missing fraction above the threshold
+    missing_col = np.where(missing_series > missing_thresh)
+    # List of indices is located in the 0 element of missing_col array
+    missing_col = missing_col[0]
+
+    print('%d columns with more than %0.2f missing values:\n\n' % (len(missing_col), missing_thresh))
+    for i in range(len(missing_col)):
+        print('Column # %d with %0.2f missing values.\n' % (missing_col[i], missing_series[missing_col[i]]))
+
+    return missing_col
+
+
+def single_value(data):
+    # type: (ndarray) -> ndarray
+
+    """
+    Find and print columns that have only a single value.
+        -----------
+        Parameters:
+            data: numpy ndarray, shape = [n_samples, n_features]
+                    observations and features
+        -----------
+        Returns:
+            A list containing the indices of columns with only a single value
+    """
+
+    single_unique = []
+    for i in range(data.shape[1]):
+        # finding the unique value of each column and if the length of that equals 1 means
+        # that feature has only one single value, then save the index of that column in single_unique
+        if len(np.unique(data[:, i])) == 1:
+            single_unique += [i]
+
+    single_unique = np.array(single_unique)
+    print('%d columns with a single unique value:\n\n' % len(single_unique))
+    for i in range(len(single_unique)):
+        print('Column # %d with a single value of %s \n' % (single_unique[i], data[1, single_unique[i]]))
+
+    return np.array(single_unique)
+
+
+def collinear(train_data, corr_thresh=0.9):
+    # type: (ndarray, float) -> ndarray
+
+    """
+    Find and print columns with correlation coefficient value more than specified value
+        -----------
+        Parameters:
+            train_data: numpy ndarray, shape = [n_samples, n_features]
+                    observations and features
+            corr_thresh: float between 0 and 1
+                    Pre-defined threshold set by the user, default value = 0.9
+        -----------
+        Returns:
+            A list of tuples containing a pair of column indices with
+            correlation coefficient value more than specified value
+    """
+
+    # creating a matrix for correlation coefficient
+    corr_matrix = np.corrcoef(train_data, rowvar=False)
+
+    # finding pairs with corrcoef more than corr_thresh and save them as a tuple in corr_pairs
+    corr_pairs = []
+    for i in range(corr_matrix.shape[0]):
+        for j in range(i + 1, corr_matrix.shape[0]):
+            if abs(corr_matrix[i, j]) > corr_thresh:
+                corr_pairs += [(i, j)]
+
+    print('%d features with a correlation greater than %0.2f:\n\n' % (len(np.unique(corr_pairs)), corr_thresh))
```

Comment:
I think this change doesn't actually change the issue & output at all, because all the pairs were already unique.  I think that the simplest reasonable fix is to *instead simply change the text* to read something like ```'%d feature pairs with...'```

Something to consider for the future that I think would be even better but a bit trickier (and not something I'd expect you to tackle now), would be to report strongly correlated groups of features.   Suppose that features a, b, c, and d were all strongly correlated to each other, and that feature e was strongly correlated to feature a but not to the other features.  The current approach would report 7 pairs of strongly correlated features, but how does the user then decide that to do with that?  If you consider the features to be vertices in a graph, with edges between those features that had pair-wise correlation above the threshold, and determine [the cliques of the graph](https://en.wikipedia.org/wiki/Clique_(graph_theory)), then the strongly correlated groups would be the maximal cliques (you'd only report those with size larger than 1 vertex), and a reasonable feature reduction approach would be to have a single proxy feature for each maximal clique (or perhaps not each, just enough to make a clique cover).   In this example with features a-e, rather than getting a listing of 7 feature pairs, the user would be told
```
With correlation threshold X%,
Features a, b, c, d form  a strongly correlated group
Features a, e form a strongly correlated group
```